cli/command/image/build.go

```diff
@@ -247,6 +253,66 @@ func runBuild(dockerCli *command.DockerCli, options buildOptions) error {
 		buildCtx = replaceDockerfileTarWrapper(ctx, buildCtx, relDockerfile, translator, &resolvedTags)
 	}
 
+	// write ctx tar to disk, to make it readable for Image Reference extraction
+	originCtx := buildCtx
+	defer originCtx.Close()
+	readbackTar, err := ioutil.TempFile("", "docker-ctx")
+	defer os.Remove(readbackTar.Name())
+	defer readbackTar.Close()
+	if err != nil {
+		return err
+	}
+	_, err = io.Copy(readbackTar, buildCtx)
+	if err != nil {
+		return err
+	}
+	_, err = readbackTar.Seek(0, os.SEEK_SET)
+	if err != nil {
+		return err
+	}
+
+	// read dockerfile and extract registry information
+
+	referencedRegs := make(map[string]bool)
+
+	observer := func(ctx context.Context, ref reference.Named) {
+		repoInfo, err := registry.ParseRepositoryInfo(ref)
+		if err != nil {
+			// can't get repo info, ignore
+			return
+		}
+		if repoInfo != nil && repoInfo.Index != nil {
+			referencedRegs[registry.GetAuthConfigKey(repoInfo.Index)] = true
+		}
+	}
+
+	tarReader := tar.NewReader(readbackTar)
+
+	for {
+		hdr, err := tarReader.Next()
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return err
+		}
+
+		if hdr.Name == relDockerfile {
+			content := io.Reader(tarReader)
+			_, _, err = rewriteDockerfileFrom(ctx, content, nil, observer)
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	_, err = readbackTar.Seek(0, os.SEEK_SET)
+	if err != nil {
+		return err
+	}
+
+	buildCtx = readbackTar
```

Comment:
The block above (~60 lines) is being added to an already massive function. 

It seems to be pretty independent, could we make this a separate function (or even a couple functions)?